convert-hf-to-gguf.py

```diff
@@ -1329,10 +1342,12 @@ def modify_tensors(self, data_torch: Tensor, name: str, bid: int | None) -> Iter
         n_head = self.hparams["num_attention_heads"]
         n_kv_head = self.hparams.get("num_key_value_heads")
 
-        if name.endswith("q_proj.weight"):
-            data_torch = LlamaModel.permute(data_torch, n_head, n_head)
-        if name.endswith("k_proj.weight"):
-            data_torch = LlamaModel.permute(data_torch, n_head, n_kv_head)
+        # Skip for granite models
+        if self.hparams.get("vocab_size", 32000) != 49152:
+            if name.endswith("q_proj.weight"):
+                data_torch = LlamaModel.permute(data_torch, n_head, n_head)
+            if name.endswith("k_proj.weight"):
+                data_torch = LlamaModel.permute(data_torch, n_head, n_kv_head)
```

Comment:
I think we can avoid adding the rope type parameter all together, by permuting the Q, K attention tensors in the correct way here. I don't have an example code unfortunately, so we need to figure out how to do it. The only difference between RoPE NORM and NEOX is that in the former we rotate the pairs `(x[2*i + 0], x[2*i + 1]`, while in the latter we rotate `(x[i], x[i + n_rot/2])`. So it's a matter of reordering the rows in each head in the correct way to make the RoPE type to be NORM - as all other LLaMA-based models

Reply:
Even though the output of `Q = rope(q)` and `K = rope(k)` would not be in the same order, it should still work because we compute `KQ = K @ Q` which is invariant to how the data in the heads is reordered - as long as it is reordered in the same way in both `K` and `Q`

I could be missing something though - not 100% confident in this. If you think it won't work, we can probably do the rope type thing, but I really prefer to find a way to avoid it

Reply:
is it something that could be changed later?

I am not confident either that it is not possible, I've spent a few hours on it and I've not been successful so far

Reply:
I gave this a try. Only the first `n_dims` elements of each rows should be re-ordered.

https://github.com/ggerganov/llama.cpp/blob/95f84d5ce8b449a9b16009434aca800df504a02e/ggml.c#L14418

```python3
    @staticmethod
    def permute_neox_rope(weights: Tensor, rot_dim: int) -> Tensor:
        orig_shape = weights.shape
        assert orig_shape[-1] % rot_dim == 0
        # reorder the first rot_dim elements of each row
        weights = weights.reshape((-1 , weights.shape[-1] // rot_dim, rot_dim))
        weights[:, 0, :] = weights[:, 0, :].reshape((-1, 2, rot_dim // 2)).mT.contiguous().reshape((-1, rot_dim))
        return weights.reshape((orig_shape))
```

It seems to partially work, but the output is still wrong, because in RoPE NEOX, it's only the first `rot_dim` elements per row that are roped, while in RoPE NORM, all of them are.

So it's not simply a re-ordering of elements that is necessary, unfortunately. The rope type is needed, it seems.

Reply:
Ah, true. So technically there is a difference between the 2 modes. But in this case `n_rot == n_embd_head == 80` so it should make a difference (I'm testing with `granite-3b-code-base`)

> # reorder the first rot_dim elements of each row

I'm not used to the Python syntax, so not sure but note that we have to reorder the *rows* in the weight tensors, not the elements in the rows. Later when we multiply these tensors with the hidden state, the rows "become" the elements in the row that we rope. Does that make sense?

Reply:
ah, if you are fine that I apply it directly on top of my patch then I can do.

I was thinking about you owning the ownership for the commit, since you came up with the code

Reply:
PR updated

Reply:
Nice work, thanks for looking into this!